### ISSUE TITLE
ANSI colors regression compared to Notebook

### ISSUE THREAD

**Issue by mgeier** [jupyterlab/jupyterlab#3773]
I've spent a considerable amount of work on the handling of ANSI colors in the Jupyter Notebook, mainly in https://github.com/jupyter/notebook/blob/master/notebook/static/base/js/utils.js.

It looks like when switching to JupyterLab, this was mostly for the bin.

At least my choice of colors has been taken in #2457, but some CSS names have been changed e.g. from `ansi-yellow-intense-fg` to `ansi-bright-yellow-fg`. This might affect people who want to use their Notebook styles for JupyterLab (I don't know if that's at all relevant, though).

More importantly, there seems to be no support for bold fonts. The ANSI escape sequences for "bold" only switch to the "bright" (or rather "intense") color (which is good), but don't switch to a bold font.

ANSI "underline" and "inverse" don't work either. Granted, support for those was only added to the Notebook quite recently.

The extended colors don't seem to work either (I tried it on a local installation and on binder), which is I guess due to a myriad of CSS definitions like `ansi-palette-201-fg` that don't seem to be available (and honestly, those hundreds of CSS classes seem a bit over-the-top to me).

Finally, 24-bit colors are not showing up. I don't know how this is supposed to work, the generated HTML looks like this:

```html
<span class="ansi-truecolor-fg ansi-truecolor-bg" data-ansi-truecolor-fg="25, 0, 230" data-ansi-truecolor-bg="230, 0, 25">
```

All of the above works fine in the latest release of the Notebook.

So what about dumping ansi_up and just using the good old code from the Notebook instead?

---

Here are two test notebooks:

* https://github.com/jupyter/notebook/blob/master/tools/tests/ANSI%20Test.ipynb
  * binder Notebook: https://mybinder.org/v2/gh/jupyter/notebook/master?filepath=tools%2Ftests%2FANSI%20Test.ipynb
  * binder JupyterLab: https://mybinder.org/v2/gh/jupyter/notebook/master?urlpath=lab (navigate to `tools/tests/ANSI Test.ipynb`)
* http://nbsphinx.readthedocs.io/en/latest/code-cells.html#ANSI-Colors
  * binder Notebook: https://mybinder.org/v2/gh/spatialaudio/nbsphinx/master?filepath=doc/code-cells.ipynb
  * binder JupyterLab: https://mybinder.org/v2/gh/spatialaudio/nbsphinx/master?urlpath=lab (navigate to `doc/code-cells.ipynb` and scroll down a bit)

Note that the Notebook version currently used on binder doesn't support proper ANSI "inverse" yet (but the latest Notebook release does), nor does the latest `nbconvert` (but there is a hopefully soon-to-be-merged PR: https://github.com/jupyter/nbconvert/pull/696). Also, the rendering on Github isn't quite ideal, either.

**Comment by ellisonbg** [comment]
@mgeier I don't think anyone working on lab has a strong usage of, nor opinions about ANSI colors. We would love to have your help (or at least advice) getting them in better shape in JupyterLab - especially as it looks like you have a lot of relevant experience. The regressions compared to the classic notebook in the ANSI colors are not deliberate - hopefully with a small bit of work, just temporary.

The test notebooks you have provided will be super helpful in bringing everything over.

I don't really have any opinion about ansi-up, versus the classic notebook's code.

Do you want to have a go at a PR, or do you think there are any decisions to make before you or someone dives into the code?

**Comment by rgbkrk** [comment]
Ooh I want to watch over this issue. :D

**Comment by mgeier** [comment]
> do you think there are any decisions to make before you or someone dives into the code?

I don't know if I'm aware of all the options.

One option would be to try to fix ansi_up, but I think there is too much wrong with it for that to be feasible.

Another option would probably be to use xterm.js, since that's a dependency already.
Did you ever consider to use xterm.js for output cells that potentially contain ANSI escape sequences?
This would probably also handle carriage return and backspace characters automatically.
The support for ANSI colors seems to be quite good. There's just a minor problem with "inverse" (which is probably fixed in https://github.com/xtermjs/xterm.js/pull/996) and 32-bit colors are just "emulated" with 256 colors (https://github.com/xtermjs/xterm.js/issues/484), but that's IMHO not a big deal.
But probably using xterm.js many times per notebook is overkill?
And I don't know what the performance implications would be.

We could try to find yet another package that's more correct than ansi_up but probably less complex than xterm.js?

Or we could use the old Notebook code.
Since JupyterLab depends on the Notebook (it does, doesn't it?), is there a way to just call those JavaScript functions?
If not, where to put the code?
Would we have to port it to TypeScript?

Are there any other options that I missed?

> Do you want to have a go at a PR

I don't have a lot of experience with JavaScript and I know nothing about TypeScript.
Nor do I know anything about the JupyterLab code. I couldn't even manage to make a development installation yet.
So it would probably be better if somebody else does it.

However, if nobody else wants to do it, I can try to tackle it, but it will probably take a long time and I'll need a lot of help ...

**Comment by blink1073** [comment]
Hi @mgeier, thanks for engaging on this subject!  It is a shame to hear that `ansi_up` isn't doing as much as we thought it did.  Our preference has been to use existing `npm` packages where possible and port the classic notebook code where needed.  I see two options moving forward:

- Port the ansi handling code from the classic notebook
- Use xterm 3.0 for console text iff it contains ansi escape sequences.  

The reason I say xterm 3.0 is that they switched to canvas based drawing.  I suspect that it would be overkill and take longer for the initial render when we don't need it, but it is worth trying as an experiment I think.

**Comment by mgeier** [comment]
Yes @blink1073, trying xterm.js sounds like an interesting (and possibly doomed) experiment. Sadly, I don't have the abilities to do that.

There is another concern: copying text from the xterm.js area could be problematic, at least it is currently in the JupyterLab Terminal.
Probably this was improved in xterm.js 3.0?

What about updating the JupyterLab Terminal to the latest xterm.js version?

**Comment by blink1073** [comment]
Yep, xterm.js 3.0 will be in the [next release](https://github.com/jupyterlab/jupyterlab/pull/3553).

**Comment by jasongrout** [comment]
@mgeier, is this still an issue with the current JupyterLab 0.34?

**Comment by mgeier** [comment]
@jasongrout 

> is this still an issue with the current JupyterLab 0.34?

AFAICT, "underline" and "inverse" don't work at all, and using "bold" doesn't switch to the "bright"/"intense" color (see https://github.com/jupyterlab/jupyterlab/pull/4230#issuecomment-379333366).
The colors, both indexed and 24-bit RGB, seem to work fine now.

The current Classic Notebook and `nbconvert` work correctly, for example you can try this:

```python
print('AB\x1b[43mCD\x1b[35mEF\x1b[1mGH\x1b[4mIJ\x1b[7m'
      'KL\x1b[49mMN\x1b[39mOP\x1b[22mQR\x1b[24mST\x1b[27mUV')
```

The result should look something like this:

![grafik](https://user-images.githubusercontent.com/705404/45584570-73cb1000-b8d6-11e8-8ab1-bb504a9bfdf7.png)

Note that between `F` and `G`, when switching to "bold", the color also changes to the "intense" variant.
Between `H` and `I`, obviously the underlined section begins.
And between `J` and `K`, foreground and background are inverted.

**Comment by mgeier** [comment]
I've created #5336 as proof of concept for what it would take to replace `ansi_up` with my code from the Classic Notebook.

Any opinions?
